freeipa/src/main/java/com/sequenceiq/freeipa/service/image/FreeIpaImageProvider.java

```diff
@@ -0,0 +1,167 @@
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(FreeIpaImageProvider.class);
+
+    private static final Pattern VERSION_PATTERN = Pattern.compile("^([0-9]+\\.[0-9]+\\.[0-9]+-(dev\\.|rc\\.|[b]))[0-9]+$");
+
+    @Inject
+    private ImageCatalogProvider imageCatalogProvider;
+
+    @Value("${freeipa.image.catalog.url}")
+    private String defaultCatalogUrl;
+
+    @Value("${freeipa.image.catalog.default.os}")
+    private String defaultOs;
+
+    @Value("${info.app.version:}")
+    private String freeIpaVersion;
+
+    @Override
+    public Optional<ImageWrapper> getImage(ImageSettingsRequest imageSettings, String region, String platform) {
+        String imageId = imageSettings.getId();
+        String catalogUrl = Objects.nonNull(imageSettings.getCatalog()) ? imageSettings.getCatalog() : defaultCatalogUrl;
+        String imageOs = Objects.nonNull(imageSettings.getOs()) ? imageSettings.getOs() : defaultOs;
+
+        ImageCatalog cachedImageCatalog = imageCatalogProvider.getImageCatalog(catalogUrl);
+        return findImageForAppVersion(region, platform, imageId, imageOs, cachedImageCatalog)
+                .or(() -> retryAfterEvictingCache(region, platform, imageId, catalogUrl, imageOs))
+                .map(i -> new ImageWrapper(i, catalogUrl, null));
+    }
+
+    private Optional<Image> findImageForAppVersion(String region, String platform, String imageId, String imageOs, ImageCatalog catalog) {
+        List<FreeIpaVersions> versions = filterFreeIpaVersionsByAppVersion(catalog.getVersions().getFreeIpaVersions());
+        List<Image> compatibleImages = findImage(imageId, imageOs, catalog.getImages().getFreeipaImages(), region, platform);
+        LOGGER.debug("[{}] compatible images found, by the following parameters: imageId: {}, imageOs: {}, region: {}, platform: {}",
+                compatibleImages.size(), imageId, imageOs, region, platform);
+
+        return findImageInDefaults(versions, compatibleImages)
+                .or(() -> findImageByApplicationVersion(versions, compatibleImages))
+                .or(() -> findMostRecentImage(compatibleImages));
+    }
+
+    private List<Image> findImage(String imageId, String imageOs, List<Image> images, String region, String platform) {
+        if (Objects.nonNull(imageOs) && !imageOs.isEmpty()) {
+            images = filterImages(images, imageOs, platform, region);
+        }
+        if (Objects.nonNull(imageId) && !imageId.isEmpty()) {
+            return images.stream()
+                    .filter(img -> img.getImageSetsByProvider().containsKey(platform) && img.getImageSetsByProvider().get(platform).containsKey(region))
+                    .filter(img -> img.getImageSetsByProvider().get(platform).get(region).equalsIgnoreCase(imageId))
+                    .collect(Collectors.toList());
+        } else {
+            return images.stream().filter(image -> image.getImageSetsByProvider().containsKey(platform))
```

Comment:
The if filters for region too (not just platform), do we not need that here?

Reply:
Copy-paste from ImageService